libsolutil/StringUtils.h

```diff
@@ -101,109 +101,113 @@ std::string joinHumanReadablePrefixed
 }
 
-/// Same as @ref formatNumberReadable but only for unsigned numbers
+/// Formats large numbers to be easily readable by humans.
+/// Returns decimal representation for smaller numbers; hex for large numbers.
+/// "Special" numbers, powers-of-two and powers-of-two minus 1, are returned in
+/// formulaic form like 0x01 * 2**24 - 1.
+/// @a T can be any integer variable, will typically be u160, u256 or bigint.
+/// @param _value to be formatted
+/// @param _useTruncation if true, internal truncation is also applied,
+/// like  0x5555...{+56 more}...5555
+/// @example formatNumberReadable((u256)0x7ffffff) = "0x08 * 2**24"
+/// @example formatNumberReadable(-57896044618658097711785492504343953926634992332820282019728792003956564819968) = -2**255
 template <class T>
-inline std::string formatUnsignedNumberReadable (
+inline std::string formatNumberReadable(
 	T const& _value,
 	bool _useTruncation = false
 )
 {
 	static_assert(
-		std::is_same<bigint, T>::value || !std::numeric_limits<T>::is_signed,
-		"only unsigned types or bigint supported"
-	); //bigint does not carry sign bit on shift
+		std::numeric_limits<T>::is_integer,
+		"only integer numbers are supported"
+	);
+
+	bool isNegative = _value < 0;
+	bigint temp;
+	bigint signedValue = isNegative ? (bigint(-1) * _value) : bigint(_value);
+	std::string sign = isNegative ? "-" : "";
 
 	// smaller numbers return as decimal
-	if (_value <= 0x1000000)
-		return _value.str();
+	if (signedValue <= 0x1000000)
+		return sign + signedValue.str();
 
 	HexCase hexcase = HexCase::Mixed;
-	HexPrefix prefix = HexPrefix::Add;
+	HexPrefix hexprefix = HexPrefix::Add;
 
-	// when multiple trailing zero bytes, format as N * 2**x
+	// when multiple trailing zero bytes, format as N * 2**x (if positive)
+	// or -N * 2**x (if negative)
 	int i = 0;
-	T v = _value;
-	for (; (v & 0xff) == 0; v >>= 8)
+	for (temp = signedValue; (temp & 0xff) == 0; temp >>= 8)
 		++i;
 	if (i > 2)
 	{
 		// 0x100 yields 2**8 (N is 1 and redundant)
-		if (v == 1)
-			return "2**" + std::to_string(i * 8);
-		return toHex(toCompactBigEndian(v), prefix, hexcase) +
-			" * 2**" +
-			std::to_string(i * 8);
+		if (temp == 1)
+			return sign + "2**" + std::to_string(i * 8);
+		else if ((temp & (temp-1)) == 0)
+		{
+			int j = 0;
+			for (; (temp & 0x1) == 0; temp >>= 1)
+				j++;
+			return sign + "2**" + std::to_string(i * 8 + j);
+		}
+		else
+			return sign +
+				toHex(toCompactBigEndian(temp), hexprefix, hexcase) +
+				" * 2**" +
+				std::to_string(i * 8);
 	}
 
-	// when multiple trailing FF bytes, format as N * 2**x - 1
+	// when multiple trailing FF bytes, format as N * 2**x - 1 (if positive)
+	// or -N * 2**x + 1 (if negative)
 	i = 0;
-	for (v = _value; (v & 0xff) == 0xff; v >>= 8)
+	for (temp = signedValue; (temp & 0xff) == 0xff; temp >>= 8)
 		++i;
 	if (i > 2)
 	{
+		std::string suffix = isNegative ? " + 1" : " - 1";
+
 		// 0xFF yields 2**8 - 1 (v is 0 in that case)
```

Comment:
```suggestion
		// 0xFF yields 2**8 - 1 (temp is 0 in that case)
```
Also, I guess we could remove `(temp is 0 in that case)`.

Reply:
It could help other programmers update the code.